File: elixir/protein-translation/lib/protein_translation.ex
defmodule ProteinTranslation do
  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {:ok, list(String.t())} | {:error, String.t()}
  def of_rna(rna) do
    translation =
      ~r/.{1,3}/
      |> Regex.scan(rna)
      |> List.flatten()
      |> Enum.map(&protein_for/1)
      |> Enum.reduce(
        %{list: [], stopped: false},
        fn p, %{list: list, stopped: stopped?} = translation ->
          cond do
            stopped? -> translation
            p == "STOP" -> %{list: list, stopped: true}
            true -> %{list: list ++ [p], stopped: false}
          end
        end
      )
      |> Map.get(:list)

    if translation |> Enum.any?(&is_nil/1) do
      {:error, "invalid RNA"}
    else
      {:ok, translation}
    end
  end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def of_codon(codon) do
    case protein_for(codon) do
      nil -> {:error, "invalid codon"}
      protein -> {:ok, protein}
    end
  end

  @spec protein_for(String.t()) :: String.t() | nil
  defp protein_for(codon) do
    case codon do
      "UGU" -> "Cysteine"
      "UGC" -> "Cysteine"
      "UUA" -> "Leucine"
      "UUG" -> "Leucine"
      "AUG" -> "Methionine"
      "UUU" -> "Phenylalanine"
      "UUC" -> "Phenylalanine"
      "UCU" -> "Serine"
      "UCC" -> "Serine"
      "UCA" -> "Serine"
      "UCG" -> "Serine"
      "UGG" -> "Tryptophan"
      "UAU" -> "Tyrosine"
      "UAC" -> "Tyrosine"
      "UAA" -> "STOP"
      "UAG" -> "STOP"
      "UGA" -> "STOP"
      _ -> nil
    end
  end
end

File: cpp/protein-translation/protein_translation.cpp
#include "protein_translation.h"



namespace protein_translation {

CodificationMap default_codification_map() {
    return {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"},
        {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},
        {"UUG", "Leucine"},
        {"UCU", "Serine"},
        {"UCC", "Serine"},
        {"UCA", "Serine"},
        {"UCG", "Serine"},
        {"UAU", "Tyrosine"},
        {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},
        {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"},
        {"UAG", "STOP"},
        {"UGA", "STOP"}
    };
}

template< typename T>
bool contains( const std::vector<T>& vec, T element ) {

    const int members_number = vec.size();

    for( int i = 0; i < members_number; i++ ) {

        if( vec.at( i ) == element )
            return true;

    }
    
    return false;

}

Protein protein( const Codon codon, const CodificationMap& codification_map ) {

    if( codification_map.find( codon ) != codification_map.end() )
        return codification_map.at( codon );
    
    return "STOP";

}

int codon_index( const int nth_codon ) {

    const int codon_len = 3;

    return (nth_codon - 1) * codon_len;

}

std::vector<Codon> split_in_codons( const RNA& rna ) {

    const int codon_len = 3;
    const int number_of_codons = std::floor( rna.size() / codon_len );
    std::vector<Codon> codons;

    for( int i = 1; i <= number_of_codons; i++ ) {

        const int codon_start = codon_index( i );
        const Codon codon = rna.substr( codon_start, codon_len );
        
        codons.push_back( codon );

    }

    return codons;

}

std::vector<Protein> proteins( const std::vector<Codon>& codons ) {

    const int number_of_codons = codons.size();
    std::vector<Protein> codifieds;
    const CodificationMap codification_map = default_codification_map();

    for( int i = 0; i < number_of_codons; i++ ) {

        const Codon current_codon = codons.at( i );
        const Protein codefied = protein( current_codon, codification_map );

        if( codefied == "STOP" )
            break;

        if( codefied != "" )
            codifieds.push_back( codefied );

    }

    return codifieds;

}

std::vector<Protein> proteins( const RNA rna ) {

    return proteins( split_in_codons( rna ) );

}



}  // namespace protein_translation
